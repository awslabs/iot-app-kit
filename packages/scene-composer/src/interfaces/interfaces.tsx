/**
 * This file contains public interfaces of the scene content the SceneComposer exposes.
 * It is different from the serialization format from the GLTF Toolkit package.
 * These interfaces are used to interact with the SceneComposer instance at
 * runtime, thus the data structure is slightly different from the GLTF Toolkit.
 */

import { ReactElement } from 'react';

import { ISceneNodeInternal } from '../store/Store';

import { ISceneComponent, KnownComponentType } from './components';
import { IRuleBasedMap, ITransform, ITransformConstraint, Vector3 } from './dataTypes';

/************************************************
 * Scene document, entity, component models
 ************************************************/

export interface ISceneDocumentSnapshot {
  /**
   * Serialize the current document snapshot into a document string with the
   * given format version.
   *
   * Note that this interface is not optimized for very large scene, which
   * we may introduce new APIs later to use stream API to handle the output.
   *
   * TODO: do we want to allow updating the document version?
   * [Option 1]: Externalize the document version instea of keeping the doucment version in the document itself.
   * [Option 2]: Allow passing in a parameter to set the document version.
   * [Option 3]: Automatically increment the document version by 1 based on the original input document version.
   *
   * We'll go with option 1 for now for simplicity.
   *
   * TODO: what should we return when the scene is empty?
   * [Option 1]: Return a falsy value, such as empty string.
   * [Option 2]: Return a valid scene with no entity.
   *
   * We'll go with option 2 for now.
   *
   * @param specVersion - version of the document format
   */
  serialize(specVersion: string, ...stringifyArgs): string;
}

export interface ISceneDocument {
  nodeMap: Record<string, ISceneNode>;
  ruleMap: Record<string, IRuleBasedMap>;
  rootNodeRefs: string[];
  unit?: string;
  version: string;
  properties?: Partial<Record<KnownSceneProperty, any>>;
  specVersion?: string;
}

export interface ISceneNode {
  // ref is optional when used in APIs to create nodes
  ref?: string;
  name?: string;
  transform?: ITransform;
  transformConstraint?: ITransformConstraint;
  components?: ISceneComponent[];
  parentRef?: string;
  childRefs?: string[];
  properties?: Partial<Record<'alwaysVisible' | 'matterportId' | 'hiddenWhileImmersive', boolean | string>>;
}

export enum KnownSceneProperty {
  BaseUrl = 'baseUrl',
  EnvironmentPreset = 'environmentPreset',
  DataBindingConfig = 'dataBindingConfig',
  ComponentSettings = 'componentSettings',
  MatterportModelId = 'matterportModelId',
  LayerIds = 'layerIds',
  SceneRootEntityId = 'sceneRootEntityId',
  TagCustomColors = 'tagCustomColors',
  FogSettings = 'fogSettings',
}

/************************************************
 * Scene Composer Control APIs
 ************************************************/

export type CameraControlMode = 'transition' | 'teleport';
export type CameraControlsType = 'orbit' | 'pan' | 'immersive' | 'pointerLock';
export type TransformControlMode = 'translate' | 'rotate' | 'scale';
export enum CameraViewAxisValues {
  Front = 'front',
  Back = 'back',
  Left = 'left',
  Right = 'right',
  Top = 'top',
  Bottom = 'bottom',
}

// an object ref or a precise camera setting
export type FixedCameraTarget = { position: Vector3; target: Vector3 };
export type NamedCameraTarget = string;
export type CameraTarget = NamedCameraTarget | FixedCameraTarget;

export type URIModifier = (uri: string) => string;

export type CreateErrorViewCallback = (error: Error) => ReactElement;

/************************************************
 * Scene Namespace
 ************************************************/

const IotTwinMakerNamespace = 'iottwinmaker.common';
export const IotTwinMakerNamespaceSeparator = ':';
export const IotTwinMakerIconNamespace = `${IotTwinMakerNamespace}.icon`;
export const IotTwinMakerColorNamespace = `${IotTwinMakerNamespace}.color`;
export const IotTwinMakerNumberNamespace = `${IotTwinMakerNamespace}.number`;
export const IotTwinMakerOpacityNamespace = `${IotTwinMakerNamespace}.opacity`;
/************************************************
 * Scene Resource
 ************************************************/

export enum SceneResourceType {
  Icon = 'Icon',
  Color = 'Color',
  Number = 'Number',
  Opacity = 'Opacity',
}

export interface SceneResourceInfo {
  type: SceneResourceType;
  value: string;
}

export interface AddingWidgetInfo {
  type?: KnownComponentType;
  node: ISceneNodeInternal;
}

export interface MeshStyle {
  color?: string | number;
  opacity?: number;
  transparent?: boolean;
}

export interface StyleTarget {
  dataBindingContext: unknown;
  style: MeshStyle;
}

export interface IFogSettings {
  colorHex: number;
  near: number;
  far: number;
}
