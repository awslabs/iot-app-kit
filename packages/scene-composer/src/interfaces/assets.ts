export enum AssetType {
  BIN = 'BIN',
  GLB = 'GLB',
  GLTF = 'GLTF',
  JPG = 'JPG',
  JPEG = 'JPEG',
  MP3 = 'MP3',
  MP4 = 'MP4',
  PDF = 'PDF',
  PNG = 'PNG',
  TILES_3D = 'Tiles3D',
}

export const ModelFileTypeList: AssetType[] = [AssetType.GLB, AssetType.GLTF, AssetType.TILES_3D];
export const TextureFileTypeList: AssetType[] = [AssetType.JPG, AssetType.JPEG, AssetType.PNG];
