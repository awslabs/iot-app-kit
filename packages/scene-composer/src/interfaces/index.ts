export * from './assets';
export * from './interfaces';
export * from './components';
export * from './dataTypes';
export * from './dataBinding';
export * from './metricRecorder';
export * from './feature';
export * from './sceneViewer';
export * from './componentSettings';
