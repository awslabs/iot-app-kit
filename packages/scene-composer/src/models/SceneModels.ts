/**
 * This file contains the model of the scene file stored in database.
 * Changes to this file should be backward compatible.
 */

export type KeyValuePair = { [key: string]: unknown };
export type UUID = string;
export type Vector3 = [number, number, number];
export type Vector2 = [number, number];
export type Uri = string;

export type DistanceUnit =
  | 'millimeters'
  | 'centimeters'
  | 'decimeters'
  | 'meters'
  | 'kilometers'
  | 'inches'
  | 'feet'
  | 'yards'
  | 'miles';

export type Color = number | string;

export interface GeoLocation {
  longitude: number;
  latitude: number;
  altitude?: number;
}

export interface RuleStatement {
  expression: string;
  target: string;
}

export interface RuleBasedMap {
  statements: Array<RuleStatement>;
}

export interface ValueDataBinding {
  dataBindingContext: unknown;
}

export interface NavLink {
  destination?: string;
  params?: Record<string, string>;
}

export interface Scene {
  specVersion: string;
  version: string;
  unit?: DistanceUnit;
  nodes: Array<Node>;
  rootNodeIndexes: Array<number>;
  properties?: KeyValuePair;
  rules?: Record<string, RuleBasedMap>;
  cameras?: Array<Camera>;
  defaultCameraIndex?: number;
}

export enum CameraType {
  Perspective = 'Perspective',
  Orthographic = 'Orthographic',
}

export interface Camera {
  cameraType: CameraType;
  fov?: number;
  near?: number;
  far?: number;
}

export enum LightType {
  Ambient = 'Ambient',
  Directional = 'Directional',
  Hemisphere = 'Hemisphere',
  Point = 'Point',
}

export interface Transform {
  position: Vector3;
  rotation?: Vector3; // euler angle
  scale?: Vector3;
}

export interface TransformConstraint {
  snapToFloor?: boolean;
}

export interface Node {
  name?: string;
  transform: Transform;
  transformConstraint?: TransformConstraint;
  components?: Array<Component.IComponent>;
  children?: Array<number>;
  properties?: Record<string, any>;
}

export enum ModelType {
  GLB = 'GLB',
  GLTF = 'GLTF',
  Tiles3D = 'Tiles3D', // 3D Tiles
  Environment = 'Environment', // Environment for external sources
}

export namespace Component {
  export enum Type {
    Camera = 'Camera',
    Light = 'Light',

    ModelRef = 'ModelRef',
    SubModelRef = 'SubModelRef',
    Tag = 'Tag',
    ModelShader = 'ModelShader',
    OpacityFilter = 'OpacityFilter',
    MotionIndicator = 'MotionIndicator',
    Space = 'Space',
    DataOverlay = 'DataOverlay',
  }

  export interface IComponent {
    type: Type | string;
  }

  export interface IDataBindingRuleMap {
    valueDataBinding?: ValueDataBinding;
    ruleBasedMapId?: string;
  }

  export interface Space extends IComponent {
    spaceId: string;
    parentSpaceEntityId: string;
    bounds: Vector3;
    boundsOffset: Vector3;
  }

  export interface ModelRef extends IComponent {
    uri: Uri;
    modelType: ModelType;
    unitOfMeasure?: DistanceUnit;
    localScale?: Vector3;
    castShadow?: boolean;
    receiveShadow?: boolean;
  }

  export interface SubModelRef extends IComponent {
    parentRef: string;
    selector: string | number;
  }

  export interface Camera extends IComponent {
    cameraIndex: number;
  }

  export interface Tag extends IComponent, IDataBindingRuleMap {
    icon?: string;
    navLink?: NavLink;
    offset?: Vector3;
  }

  export interface ModelShader extends IComponent, IDataBindingRuleMap {}
  export interface OpacityFilter extends IComponent, IDataBindingRuleMap {}

  export interface IMotionIndicatorConfig {
    numOfRepeatInY: number;
    backgroundColorOpacity: number;
    defaultBackgroundColor?: Color;
    defaultForegroundColor?: Color;
    defaultSpeed?: number;
  }
  export type ILinearPlaneMotionIndicatorConfig = IMotionIndicatorConfig;

  export type ILinearCylinderMotionIndicatorConfig = IMotionIndicatorConfig;

  export type ICircularCylinderMotionIndicatorConfig = IMotionIndicatorConfig;

  export enum MotionIndicatorDataBindingName {
    Speed = 'speed',
    ForegroundColor = 'foregroundColor',
    BackgroundColor = 'backgroundColor',
  }

  export enum MotionIndicatorShape {
    LinearPlane = 'LinearPlane',
    LinearCylinder = 'LinearCylinder',
    CircularCylinder = 'CircularCylinder',
  }

  export interface MotionIndicator extends IComponent {
    valueDataBindings: { [key in MotionIndicatorDataBindingName]?: IDataBindingRuleMap };

    shape: MotionIndicatorShape;
    config:
      | ILinearPlaneMotionIndicatorConfig
      | ILinearCylinderMotionIndicatorConfig
      | ICircularCylinderMotionIndicatorConfig;
  }

  // Data Overlay
  export enum DataOverlayRowType {
    Markdown = 'Markdown',
  }
  export interface DataOverlayRow {
    rowType: DataOverlayRowType;
  }
  export interface DataOverlayMarkdownRow extends DataOverlayRow {
    content: string;
  }
  export enum DataOverlaySubType {
    TextAnnotation = 'TextAnnotation',
    OverlayPanel = 'OverlayPanel',
  }
  // TODO: Not supported in milestone 1
  export interface OverlayPanelConfig {
    isPinned: boolean;
  }

  export interface ValueDataBindingNamedMap {
    bindingName: string;
    valueDataBinding?: ValueDataBinding;
  }

  export interface DataOverlay extends IComponent {
    subType: DataOverlaySubType;
    dataRows: Array<DataOverlayMarkdownRow>;

    // TODO: Not supported in milestone 1
    // config?: OverlayPanelConfig;

    valueDataBindings: ValueDataBindingNamedMap[];
  }

  export interface ILightShadowSettings {
    shadowBias?: number;
    shadowCameraLeft?: number;
    shadowCameraRight?: number;
    shadowCameraTop?: number;
    shadowCameraBottom?: number;
    shadowMapSizeWidth?: number;
    shadowMapSizeHeight?: number;
  }

  export interface IDirectionalLightSettings {
    color: Color;
    intensity: number;

    castShadow: boolean;
    shadow?: ILightShadowSettings;
  }

  export interface IPointLightSettings {
    color: Color;
    intensity: number;
    // Maximum range of the light. Default is 0 (no limit).
    distance?: number;
    // The amount the light dims along the distance of the light. Default is 1.
    // For physically correct lighting, set this to 2.
    decay?: number;

    castShadow: boolean;
    shadow?: ILightShadowSettings;
  }

  export interface IAmbientLightSettings {
    color: Color;
    intensity: number;
  }

  export interface IHemisphereLightSettings {
    color: Color;
    groundColor: Color;
    intensity: number;
  }

  export interface Light extends IComponent {
    lightType: LightType;
    lightSettings: IDirectionalLightSettings | IAmbientLightSettings | IHemisphereLightSettings;
  }
}
