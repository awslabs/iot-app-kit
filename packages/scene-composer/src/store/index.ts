export * from './StoreOperations';
export * from './Store';
