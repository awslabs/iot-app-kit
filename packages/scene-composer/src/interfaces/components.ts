import * as SceneModels from '../models/SceneModels';

/************************************************
 * Components
 ************************************************/

import { DistanceUnit, INavLink, IValueDataBinding, Vector3 } from './dataTypes';

export enum KnownComponentType {
  ModelRef = 'ModelRef',
  SubModelRef = 'SubModelRef',
  Camera = 'Camera',
  Light = 'Light',

  Tag = 'Tag',
  ModelShader = 'ModelShader',
  MotionIndicator = 'MotionIndicator',
  DataOverlay = 'DataOverlay',
}

export interface ISceneComponent {
  // ref is optional when used in APIs to create components
  ref?: string;
  type: KnownComponentType | string;
}

export interface IModelRefComponent extends ISceneComponent {
  uri: string;
  modelType: string;

  unitOfMeasure?: DistanceUnit;
  localScale?: Vector3;
  castShadow?: boolean;
  receiveShadow?: boolean;
}

export interface ISubModelRefComponent extends ISceneComponent {
  parentRef?: string;
  selector: string | number;
}

export interface IAnchorComponent extends ISceneComponent {
  icon?: string;
  valueDataBinding?: IValueDataBinding;
  ruleBasedMapId?: string;
  navLink?: INavLink;
  offset?: Vector3;
}

/**
 * Default Anchor status strings
 */
export enum DefaultAnchorStatus {
  Info = 'Info',
  Warning = 'Warning',
  Error = 'Error',
  Video = 'Video',
}

/**
 * Selected Anchor string
 */
export const SelectedAnchor = 'Selected';

/**
 * Additional Tag Component Data to be given when a tag node is clicked or changed
 */
export interface ITagData {
  navLink?: INavLink;
  dataBindingContext?: unknown;
}

/**
 * Type that can be represented by different additional component data types such as ITagData | IFutureComponentData
 */
export type AdditionalComponentData = ITagData;

/**
 * Callback signature for selection of with Widgets.
 */
export interface ISelectionChangedEvent {
  componentTypes: KnownComponentType[] | string[];
  nodeRef?: string;
  additionalComponentData?: AdditionalComponentData[];
}

/**
 * Callback signature for interaction with Widgets.
 */
export interface IWidgetClickEvent {
  componentTypes: KnownComponentType[] | string[];
  nodeRef: string;
  additionalComponentData?: AdditionalComponentData[];
}

export type WidgetClickEventCallback = (event: IWidgetClickEvent) => void;

export type SelectionChangedEventCallback = (event: ISelectionChangedEvent) => void;

export interface ILightComponent extends ISceneComponent, SceneModels.Component.Light {}

export type CameraType = 'Perspective' | 'Orthographic';

export interface ICameraBasics {
  cameraType: CameraType;
  fov?: number;
  near: number;
  far: number;
  zoom: number;
}

export interface CameraSettings extends ICameraBasics {
  transform: SceneModels.Transform;
}

export interface ICameraComponent extends ISceneComponent, ICameraBasics {}

export interface IColorOverlayComponent extends ISceneComponent {
  valueDataBinding?: IValueDataBinding;
  ruleBasedMapId?: string;
}

export interface IMotionIndicatorComponent extends ISceneComponent, SceneModels.Component.MotionIndicator {}

export interface IDataOverlayComponent extends ISceneComponent, SceneModels.Component.DataOverlay {}
