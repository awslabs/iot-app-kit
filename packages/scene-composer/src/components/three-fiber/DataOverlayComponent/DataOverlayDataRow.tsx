import React, { ReactElement, useCallback, useEffect, useMemo, useState } from 'react';

import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { ITwinMakerEntityDataBindingContext } from '../../../interfaces';
import { Component } from '../../../models/SceneModels';
import { useStore } from '../../../store';
import { dataBindingValuesProvider } from '../../../utils/dataBindingUtils';
import { replaceBindingVariables } from '../../../utils/dataBindingVariableUtils';
import { ReactMarkdownWrapper } from '../../wrappers/ReactMarkdownWrapper';
import './styles.scss';

export interface DataOverlayDataRowProps {
  rowData: Component.DataOverlayRow;
  overlayType: Component.DataOverlaySubType;
  valueDataBindings: Component.ValueDataBindingNamedMap[];
}

export const DataOverlayDataRow = ({
  rowData,
  overlayType,
  valueDataBindings,
}: DataOverlayDataRowProps): ReactElement => {
  const sceneComposerId = useSceneComposerId();
  const dataInput = useStore(sceneComposerId)((state) => state.dataInput);
  const dataBindingTemplate = useStore(sceneComposerId)((state) => state.dataBindingTemplate);
  const isEditing = useStore(sceneComposerId)((state) => state.isEditing());

  const [stringContent, setStringContent] = useState<string>('');

  const bindingValuesMap: Record<string, string> = useMemo(() => {
    const result = {};
    valueDataBindings.forEach((binding) => {
      if (!binding.valueDataBinding) {
        return;
      }
      const values: Record<string, unknown> = dataBindingValuesProvider(
        dataInput,
        binding.valueDataBinding,
        dataBindingTemplate,
      );
      result[binding.bindingName] =
        values[(binding.valueDataBinding.dataBindingContext as ITwinMakerEntityDataBindingContext).propertyName];
    });

    return result;
  }, [valueDataBindings, dataInput, dataBindingTemplate]);

  const updateContentWithBindingVariables = useCallback(
    (content: string) => {
      // Show original content in editing mode
      if (isEditing) {
        setStringContent(content);
        return;
      }

      setStringContent(replaceBindingVariables(content, bindingValuesMap));
    },
    [bindingValuesMap],
  );

  useEffect(() => {
    switch (rowData.rowType) {
      case Component.DataOverlayRowType.Markdown: {
        updateContentWithBindingVariables((rowData as Component.DataOverlayMarkdownRow).content);
        break;
      }
      default:
        break;
    }
  }, [bindingValuesMap, rowData]);

  switch (rowData.rowType) {
    case Component.DataOverlayRowType.Markdown: {
      return (
        <ReactMarkdownWrapper
          className={overlayType === Component.DataOverlaySubType.TextAnnotation ? 'annotation-row' : 'panel-row'}
          content={stringContent}
        />
      );
    }
    default:
      return <></>;
  }
};
