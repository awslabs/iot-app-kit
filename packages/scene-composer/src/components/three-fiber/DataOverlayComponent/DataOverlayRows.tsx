import React, { ReactElement } from 'react';

import { Component } from '../../../models/SceneModels';
import { IDataOverlayComponentInternal } from '../../../store/internalInterfaces';

import { DataOverlayDataRow } from './DataOverlayDataRow';
import './styles.scss';

export interface DataOverlayRowsProps {
  component: IDataOverlayComponentInternal;
}

export const DataOverlayRows = ({ component }: DataOverlayRowsProps): ReactElement => {
  return (
    <div
      className={`${
        component.subType === Component.DataOverlaySubType.TextAnnotation ? 'annotation-rows' : 'panel-rows'
      }`}
    >
      {component.dataRows.map((row, index) => {
        return (
          <DataOverlayDataRow
            rowData={row}
            overlayType={component.subType}
            key={index}
            valueDataBindings={component.valueDataBindings}
          />
        );
      })}
    </div>
  );
};
