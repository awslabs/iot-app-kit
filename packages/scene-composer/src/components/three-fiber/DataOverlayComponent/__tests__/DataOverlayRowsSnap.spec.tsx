import React from 'react';
import { render } from '@testing-library/react';

import { IDataOverlayComponentInternal } from '../../../../store';
import { KnownComponentType } from '../../../../interfaces';
import { Component } from '../../../../models/SceneModels';
import { DataOverlayRows } from '../DataOverlayRows';
import { DataOverlayDataRowProps } from '../DataOverlayDataRow';

jest.mock('../DataOverlayDataRow', () => ({
  DataOverlayDataRow: (...props: [DataOverlayDataRowProps, object]) => (
    <div data-testid='data-row'>{JSON.stringify(props)}</div>
  ),
}));

describe('DataOverlayRows', () => {
  const mockComponent: IDataOverlayComponentInternal = {
    ref: 'comp-ref',
    type: KnownComponentType.DataOverlay,
    subType: Component.DataOverlaySubType.OverlayPanel,
    dataRows: [
      {
        rowType: Component.DataOverlayRowType.Markdown,
        content: 'content 1',
      },
      {
        rowType: Component.DataOverlayRowType.Markdown,
        content: 'content 2',
      },
    ],
    valueDataBindings: [],
  };

  it('should render overlay panel correctly', () => {
    const { container } = render(<DataOverlayRows component={mockComponent} />);
    expect(container).toMatchSnapshot();
  });

  it('should render text annotation correctly', () => {
    const { container } = render(
      <DataOverlayRows component={{ ...mockComponent, subType: Component.DataOverlaySubType.TextAnnotation }} />,
    );
    expect(container).toMatchSnapshot();
  });
});
