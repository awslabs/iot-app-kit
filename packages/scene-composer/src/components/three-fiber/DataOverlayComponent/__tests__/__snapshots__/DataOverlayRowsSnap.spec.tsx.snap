// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataOverlayRows should render overlay panel correctly 1`] = `
<div>
  <div
    class="panel-rows"
  >
    <div
      data-testid="data-row"
    >
      [{"rowData":{"rowType":"Markdown","content":"content 1"},"overlayType":"OverlayPanel","valueDataBindings":[]},{}]
    </div>
    <div
      data-testid="data-row"
    >
      [{"rowData":{"rowType":"Markdown","content":"content 2"},"overlayType":"OverlayPanel","valueDataBindings":[]},{}]
    </div>
  </div>
</div>
`;

exports[`DataOverlayRows should render text annotation correctly 1`] = `
<div>
  <div
    class="annotation-rows"
  >
    <div
      data-testid="data-row"
    >
      [{"rowData":{"rowType":"Markdown","content":"content 1"},"overlayType":"TextAnnotation","valueDataBindings":[]},{}]
    </div>
    <div
      data-testid="data-row"
    >
      [{"rowData":{"rowType":"Markdown","content":"content 2"},"overlayType":"TextAnnotation","valueDataBindings":[]},{}]
    </div>
  </div>
</div>
`;
